testing/snapshot_test.ts

```diff
@@ -0,0 +1,175 @@
+// Copyright 2018-2022 the Deno authors. All rights reserved. MIT license.
+import { stripColor } from "../fmt/colors.ts";
+import { ensureDir } from "../fs/mod.ts";
+import { dirname, fromFileUrl, join } from "../path/mod.ts";
+import { assert, assertInstanceOf, AssertionError, fail } from "./asserts.ts";
+import { assertSnapshot } from "./snapshot.ts";
+
+class TestClass {
+  a = 1;
+  b = 2;
+  init() {
+    this.b = 3;
+  }
+  get getA() {
+    return this.a;
+  }
+  func() {}
+}
+
+const map = new Map();
+map.set("Hello", "World!");
+map.set(() => "Hello", "World!");
+map.set(1, 2);
+
+Deno.test("Snapshot Test", async (t) => {
+  await assertSnapshot(t, { a: 1, b: 2 });
+  await assertSnapshot(t, new TestClass());
+  await assertSnapshot(t, map);
+  await assertSnapshot(t, new Set([1, 2, 3]));
+  await assertSnapshot(t, { fn() {} });
+  await assertSnapshot(t, function fn() {});
+  await assertSnapshot(t, [1, 2, 3]);
+  await assertSnapshot(t, "hello world");
+});
+
+Deno.test("Snapshot Test - step", async (t) => {
+  await assertSnapshot(t, { a: 1, b: 2 });
+  await t.step("Nested", async (t) => {
+    await assertSnapshot(t, new TestClass());
+    await assertSnapshot(t, map);
+    await t.step("Nested Nested", async (t) => {
+      await assertSnapshot(t, new Set([1, 2, 3]));
+      await assertSnapshot(t, { fn() {} });
+      await assertSnapshot(t, function fn() {});
+    });
+    await assertSnapshot(t, [1, 2, 3]);
+  });
+  await assertSnapshot(t, "hello world");
+});
+
+Deno.test("Snapshot Test - Adverse String \\ ` ${}", async (t) => {
+  await assertSnapshot(t, "\\ ` ${}");
+});
+
+Deno.test("Snapshot Test - Failed Assertion", async (t) => {
+  await t.step("Object", async (t) => {
+    try {
+      await assertSnapshot(t, [1, 2]);
+      fail("Expected snapshot not to match");
+    } catch (error) {
+      assertInstanceOf(error, AssertionError);
+      await assertSnapshot(t, stripColor(error.message).split("\n"));
+    }
+  });
+  await t.step("String", async (t) => {
+    try {
+      await assertSnapshot(t, "Hello!");
+      fail("Expected snapshot not to match");
+    } catch (error) {
+      assertInstanceOf(error, AssertionError);
+      await assertSnapshot(t, stripColor(error.message).split("\n"));
+    }
+  });
+});
+
+Deno.test("Snapshot Test - Update", async (t) => {
+  const TEMP_DIR = ".tmp";
+
+  async function runTestWithUpdateFlag(test: string, deleteTempDir = true) {
+    const testDir = dirname(fromFileUrl(import.meta.url));
+    const tempDir = join(testDir, TEMP_DIR);
+    const tempTestFileName = "test.ts";
+    const tempTestFilePath = join(tempDir, tempTestFileName);
```

Comment:
I think it's fine for first pass, but ideally we like to use `Deno.makeTempDir` for making temp directory. Filed an issue for this https://github.com/denoland/deno_std/issues/2129

Reply:
My initial implementation did use this but I refactored as it resulted in a random output path which I would have had to obfuscate in the snapshot.

What's the advantage of using this by the way?

Reply:
`Deno.makeTempDir` creates a directory under the system's default temp directory and they are automatically removed by the OS some time later. So we don't need care much about those files even if the test code failed to remove it. Currently we need to remove `testing/.tmp` manually when the test case failed to remove it for some reason, and that's a little inconvenient

Reply:
Thanks for the explanation! 🙂